Add flow_framework namespace specs

junweid62 opened the pull request:
### Description
Part of Adding missing API specs [#168](https://github.com/opensearch-project/opensearch-api-specification/issues/168)

### Issues Resolved
[#508](https://github.com/opensearch-project/opensearch-api-specification/issues/508)


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dblock commented on spec/namespaces/flow_framework.yaml:
```diff
@@ -0,0 +1,294 @@
+      parameters:
+        - $ref: '#/components/parameters/flow_framework.create::query.provision'
+        - $ref: '#/components/parameters/flow_framework.create::query.validation'
+        # eslint-disable-next-line @cspell/spellchecker
```
You can add these to `.cspell`.

junweid62 replied:
will add.

dblock commented on spec/schemas/flow_framework._common.yaml:
```diff
@@ -0,0 +1,100 @@
+paths: {}
+components:
+  schemas:
+    FlowFrameworkException:
```
Exception vs. Error below, is that intentional?

I think this sets a nice precedent for specialized errors, should we have a `flow_framework._errors.yaml`?

junweid62 replied:
I will change it to `FlowFrameworkAPIDisabledError` for consistency.

Also will change the file name to `flow_framework._errors.yaml`

dblock commented on spec/schemas/flow_framework._common.yaml:
```diff
@@ -0,0 +1,100 @@
+paths: {}
+components:
+  schemas:
+    FlowFrameworkException:
```
Does this one inherit from an OpenSearch error type?

junweid62 replied:
yes , it inherit from OpenSearchException

`FlowFrameworkException extends OpenSearchException`

dbwiddis replied:
That makes it an Exception, though.  "Error" in exception handling has a specific meaning (the program should be halted).

junweid62 replied:
https://github.com/opensearch-project/flow-framework/blob/b3f9d6514d2c0530c8bffac654fe175452673169/src/main/java/org/opensearch/flowframework/rest/RestCreateWorkflowAction.java#L113C5-L146C55
How about these, should i call it error or exception ?

dbwiddis replied:
It's interesting, the rest response includes the word "error" as an "error message" although it's an "exception".  I'll defer to @dblock here on what the appropriate spec/api terminology is here.

dblock replied:
🤷‍♂️ @nhtruong ?

nhtruong replied:
We've been calling them errors through out the Spec (i.e. ReponseError). The clients also consider non 2XX responses as errors for error handling. As far as Client-Server communication is concerned (i.e. Web API), there's no distinction between Error and Exception, in my opinion. Any non-2XX response tells the user that something has gone wrong, and we usually call it a 500 Error Code etc instead of a 500 exception code.

dblock commented on tests/default/flow_framework/workflow.yaml:
```diff
@@ -0,0 +1,59 @@
+$schema: ../../../json_schemas/test_story.schema.yaml
+description: Test flow_framework endpoints.
+version: '>=2.12'
```
```suggestion
version: '>= 2.12'
```

(I think we add a space, but check ;)

dblock reviewed (approved):
This looks great! I only have some nits, up to you if you want to address them.

dbwiddis commented on spec/namespaces/flow_framework.yaml:
```diff
@@ -0,0 +1,294 @@
+        - $ref: '#/components/parameters/flow_framework.create::query.provision'
+        - $ref: '#/components/parameters/flow_framework.create::query.validation'
+        # eslint-disable-next-line @cspell/spellchecker
+        - $ref: '#/components/parameters/flow_framework.create::query.reprovision'
```
I still think you're missing `use_case`: https://opensearch.org/docs/latest/automating-configurations/workflow-templates/

dbwiddis commented on spec/schemas/flow_framework._common.yaml:
```diff
@@ -0,0 +1,100 @@
+          properties:
+            message:
+              type: string
+              # example: "This API is disabled. To enable it, set [flow_framework.enabled] to true."
```
For my own education, is this a comment to aid review or does this have any meaning in the spec?

dbwiddis reviewed (commented):
Looks great!  Just one missing param as mentioned in the earlier review (and I'm about to file a PR adding the missing docs)

junweid62 commented:
@dblock Hi, I have addressed all the comments and made the necessary changes. Regarding the suggestion about:

```
file: 'namespaces/flow_framework.yaml',
location: 'Operation: DELETE /_plugins/_flow_framework/workflow/{workflow_id}',
message: "The 404 response must be a reference object to '#/components/responses/flow_framework.create@400'."
```
I believe that using the error name directly like this:
```
          oneOf:
            - $ref: '../schemas/flow_framework._errors.yaml#/components/schemas/BadRequestError'
            - $ref: '../schemas/flow_framework._errors.yaml#/components/schemas/ConflictError'
            - $ref: '../schemas/flow_framework._errors.yaml#/components/schemas/MissingParameterError'
            - $ref: '../schemas/flow_framework._errors.yaml#/components/schemas/ParameterConflictError'
            - $ref: '../schemas/flow_framework._errors.yaml#/components/schemas/MaxWorkflowsLimitError'
            - $ref: '../schemas/flow_framework._errors.yaml#/components/schemas/WorkflowSaveError'
```


makes the it clearer in this particular case. Please let me know your thoughts on this approach.

dbwiddis commented on tests/default/flow_framework/workflow.yaml:
```diff
@@ -0,0 +1,59 @@
+      status: 201
+    output:
+      test_workflow_id: payload.workflow_id
+  - synopsis: Get workflow.
+    id: get_flow_framework
+    path: /_plugins/_flow_framework/workflow/{workflow_id}
+    method: GET
+    parameters:
+      workflow_id: ${create_flow_framework.test_workflow_id}
+    response:
+      status: 200
```
Could probably add tests for some of the other workflow 4xx returns here:
 - leave workflow_id blank (not sure if that's possible, I'd try ending the URL `/workflow/` or `/workflow//`)
 - put a wrong workflow id

junweid62 replied:
Will add in the next pr

dbwiddis commented on tests/default/flow_framework/workflow.yaml:
```diff
@@ -0,0 +1,59 @@
+    response:
+      status: 201
+    output:
+      test_workflow_id: payload.workflow_id
```
Could add some invalid JSON here that would fail parsing for another test.

junweid62 replied:
Will add in the next pr

dbwiddis commented on tests/default/flow_framework/workflow.yaml:
```diff
@@ -0,0 +1,59 @@
+        name: test_update_work_flow
+    response:
+      status: 404
+  - synopsis: Delete workflow.
+    id: delete_flow_framework
+    path: /_plugins/_flow_framework/workflow/{workflow_id}
+    method: DELETE
+    parameters:
+      workflow_id: ${create_flow_framework.test_workflow_id}
+    response:
+      status: 200
```
Can test the wrong ID here as well.

dbwiddis reviewed (approved):
LGTM!  Good for now, I've suggested some tests you can add, but those can also come in a later PR as you iterate.

dblock commented:
> Please let me know your thoughts on this approach.

Makes sense to me. I am still figuring the best way to do these things too! @nhtruong might have other ideas.

junweid62 commented:
> > Please let me know your thoughts on this approach.
> 
> Makes sense to me. I am still figuring the best way to do these things too! @nhtruong might have other ideas.
> 
> Iterate until https://github.com/opensearch-project/opensearch-api-specification/actions/runs/10713232032/job/29722019367?pr=553 is green either way?

sure, will do

nhtruong reviewed (approved):
LGTM
Feel free to merge unless you want to add more test cases.

junweid62 commented:
@dblock Hi, I just fix the lint spec errors, hope everything goes green this time !

junweid62 commented:
@dblock Hi, got all lint spec passed locally, hope everything goes well this time!

dblock commented:
👏
